pkg/ingester/user_tsdb.go

```diff
@@ -467,3 +504,113 @@ func (u *userTSDB) releaseAppendLock(acquireState tsdbState) {
 		u.inFlightAppendsStartedBeforeForcedCompaction.Done()
 	}
 }
+
+func (u *userTSDB) ownedSeriesAndShards() (int, int) {
+	u.ownedSeriesMtx.Lock()
+	defer u.ownedSeriesMtx.Unlock()
+
+	return u.ownedSeriesCount, u.ownedSeriesShardSize
+}
+
+func (u *userTSDB) getAndClearReasonForRecomputeOwnedSeries() string {
+	return u.requiresOwnedSeriesUpdate.Swap("")
+}
+
+func (u *userTSDB) triggerRecomputeOwnedSeries(reason string) {
+	u.requiresOwnedSeriesUpdate.CompareAndSwap("", reason)
+}
+
+// recomputeOwnedSeries recomputes owned series for current token ranges, and updates both owned series and shard size.
+//
+// This method returns true, if recomputation of owned series failed multiple times due to too
+// many new series being added during the computation. If no such problem happened, this method returns false.
+//
+// This method and updateTokenRanges should be only called from the same goroutine. (ownedSeries service)
+func (u *userTSDB) recomputeOwnedSeries(shardSize int, reason string, logger log.Logger) (retry bool) {
+	retry, _ = u.recomputeOwnedSeriesWithComputeFn(shardSize, reason, logger, u.computeOwnedSeries)
+	return retry
+}
+
+const (
+	recomputeOwnedSeriesMaxAttemps    = 3
+	recomputeOwnedSeriesMaxSeriesDiff = 1000
+)
+
+func (u *userTSDB) recomputeOwnedSeriesWithComputeFn(shardSize int, reason string, logger log.Logger, compute func() int) (retry bool, _ int) {
+	start := time.Now()
+
+	prevOwnedSeriesCount, shardSizePrev := u.ownedSeriesAndShards()
+
+	reportError := false
+	attempts := 0
+	var ownedNew int
+	for attempts < recomputeOwnedSeriesMaxAttemps {
+		attempts++
+
+		ownedNew = compute()
+
+		u.ownedSeriesMtx.Lock()
+
+		// Check how many new series were added while we were computing owned series.
+		seriesDiff := u.ownedSeriesCount - prevOwnedSeriesCount
+		seriesDiffOk := seriesDiff >= 0 && seriesDiff <= recomputeOwnedSeriesMaxSeriesDiff // seriesDiff should always be >= 0, but in case it isn't, we can try again.
+		if seriesDiffOk || attempts == recomputeOwnedSeriesMaxAttemps {
```

Comment:
[nit] Generally speaking I'm not a big fan of checking for for loop stop conditions inside the loop itself. Have you considered simply changing `reportError` into `success`? `success` gets initialised to false, and switched to true once value gets updated. The for loop could simply change to `for !success && attempts < recomputeOwnedSeriesMaxAttemps`.

Reply:
Are you suggesting to move update of `u.ownedSeriesCount` and `u.ownedSeriesShardSize` outside of `for` loop? Because that would surely be nicer, but then the locking situation is more complicated. Let me try and see.

Reply:
After our chat, I've reworked the method. PTAL